ipni-datalog.md

```diff
@@ -0,0 +1,172 @@
+
+4. We also have not captured relation between `blob..left` and `block..1` in any way other than they both relate to `bag...idx`.
+
+## Proposal
+
+Leverage [IPNI] such that we could resolve `entity` that [multihash] relates to
+without having to resolve and fetch [DAG Index] first.
+
+General idea is to capture relation names in a way that it could be utilized during lookups. This also would make indexing protocol open-ended and extensible by anyone without having intermediaries like W3Up having to adopt those extensions.
+
+To illustrate this lets derive another lookup table from the same [DAG Index][index example], however in this case we will replace `key` column with pair of `attribute` (describing relation to an `entity`) and `value` which is a [multihash] related to `entity`.
+
+| entity      | attribute             | value       |
+| ----------- | --------------------- | ------------|
+| `blb..left` | `dag@0.1/shard`       | `bafy..dag` |
+| `blb..left` | `dag@0.1/shard/slice` | `block..1`  |
+| `blb..left` | `dag@0.1/shard/slice` | `block..2`  |
+| `blb..left` | `dag@0.1/shard/slice` | `block..3`  |
+| `blb..left` | `dag@0.1/shard/slice` | `block..4`  |
+| `blb..right`| `dag@0.1/shard`       | `bafy..dag` |
+| `blb..right`| `dag@0.1/shard/slice` | `block..5`  |
+| `blb..right`| `dag@0.1/shard/slice` | `block..6`  |
+| `blb..right`| `dag@0.1/shard/slice` | `block..7`  |
+| `blb..right`| `dag@0.1/shard/slice` | `block..8`  |
+
+In this variant we have virtual `key` that is derived from `attribute` and `value` components in a deterministic manner e.g. by concatenating their bytes and computing a [multihash].
+
+This means that we could find shards of he `bafy..dag` via IPNI query that resolves `toKey(['bafy..dag', 'dag@0.1/shard'])` key.
+
+If system drops `blob..left` will drop all the records associated with that entity as it would be a single IPNI advertisement.
+
+Because we get shards from single query we do not need to fetch [DAG Index] to start resolving location commitments. We could start resolving them right away and we gate IPNI with something like GraphQL we could even resolve those in a single roundtrip and cache for subsequent queries.
+
+Location commitments and blob partition info can also be indexed and advertised on [IPNI] in a very similar way.
+
+| entity   | attribute             | value        |
+| ---------|-----------------------|--------------|
+| `loc..1` | `assert/location@0.2` | `blb..left` |
+| `loc..2` | `assert/location@0.2` | `blb..left` |
+| `bl..idx`| `blob/partition@0.2`  | `blb..left` |
+| `loc..3` | `assert/location@0.2` | `blb..right`|
+| `br..idx`| `blob/partition@0.2`  | `blb..right`|
+
+## Relation to Datalog
+
+In datalog facts are triples of `[entity, attribute, value]` and you e.g. run a query like  `[?blob, 'dag@0.1/shard', 'bafy..dag']` to find all `?blob`s that have `shard` relation to `bafy..dag` which (from our example) will produce
```

Comment:
Would datalog be used as a result cache for IPNI query results? So, datalog queries would only available after getting IPNI results and then getting additional related data such as blob index and location commitments. Is that right?

Reply:
I'm not sure I follow what you're asking here. I think datalog query as a composite of multiple queries. Service that receives them can decompose them in order to query combination of local cache + IPNI, aggregate / filter results and send back to the querying client.

This creates an opportunity to reduce number of queries to IPNI by caching at several layers:

1. Query can be hashed (after some normalization to making it name agnostic) which can be used as key for a query result cache ? If we have a result we can return it without any work.
2. If cache is missed query executor decomposes query into atomic queries (where some can run concurrently and  others can only run after first set received). Each atomic query is again can be looked up in local cache by a hash. In case of cache miss query can be send to IPNI
3. After query executor performs aggregation / filtering of results it can cache it by the query key and return back to client

So in worst case scenario it would be:

1. One IPNI lookup for DAG root
2. Set of concurrent slice & commitment lookups per shard

But once engine finds responses it can cache it all the future queries would require no IPNI queries. We could also tie the cache TTL to the TTLs of all the responses that way we'll get invalidation and recompilation of joined results without having to manually deal with resource management (like invalidate some indexes when commitment expries)